src/embind/emval.js

```diff
@@ -381,67 +307,83 @@ var LibraryEmVal = {
+      reflectConstruct = function(target, args) {
+        // limited polyfill for Reflect.construct that handles variadic args and native objects, but not new.target
+        return new (target.bind.apply(target, [null].concat(args)))();
+      };
+    }
+  `,
+#else
+  $reflectConstruct: 'Reflect.construct',
+#endif
+
   _emval_get_method_caller__deps: [
-    '$emval_addMethodCaller', '$emval_lookupTypes',,
+    '$emval_addMethodCaller', '$emval_lookupTypes',
     '$makeLegalFunctionName',
+    '$reflectConstruct', '$emval_returnValue',
 #if DYNAMIC_EXECUTION
     '$newFunc',
 #endif
   ],
-  _emval_get_method_caller: (argCount, argTypes) => {
+  _emval_get_method_caller: (argCount, argTypes, asCtor) => {
     var types = emval_lookupTypes(argCount, argTypes);
     var retType = types.shift();
     argCount--; // remove the shifted off return type
 
 #if !DYNAMIC_EXECUTION
     var argN = new Array(argCount);
-    var invokerFunction = (handle, name, destructors, args) => {
+    var invokerFunction = (obj, func, destructorsRef, args) => {
       var offset = 0;
       for (var i = 0; i < argCount; ++i) {
         argN[i] = types[i]['readValueFromPointer'](args + offset);
         offset += types[i]['argPackAdvance'];
       }
-      var rv = handle[name].apply(handle, argN);
+      var rv = asCtor ? reflectConstruct(func, argN) : func.apply(obj, argN);
       for (var i = 0; i < argCount; ++i) {
         if (types[i].deleteObject) {
           types[i].deleteObject(argN[i]);
         }
       }
-      return retType['toWireType'](destructors, rv);
+      return emval_returnValue(retType, destructorsRef, rv);
     };
 #else
-    var params = ["retType"];
-    var args = [retType];
-
-    var argsList = ""; // 'arg0, arg1, arg2, ... , argN'
-    for (var i = 0; i < argCount; ++i) {
-      argsList += (i !== 0 ? ", " : "") + "arg" + i;
-      params.push("argType" + i);
-      args.push(types[i]);
-    }
-
     var signatureName = retType.name + "_$" + types.map(t => t.name).join("_") + "$";
     var functionName = makeLegalFunctionName("methodCaller_" + signatureName);
     var functionBody =
-        "return function " + functionName + "(handle, name, destructors, args) {\n";
+        "return function " + functionName + "(obj, func, destructorsRef, args) {\n";
 
     var offset = 0;
+    var argsList = asCtor ? "" : "obj"; // 'obj?, arg0, arg1, arg2, ... , argN'
+    var params = ["retType"];
+    var args = [retType];
     for (var i = 0; i < argCount; ++i) {
+        if (argsList) argsList += ", ";
+        argsList += "arg" + i;
+        params.push("argType" + i);
+        args.push(types[i]);
         functionBody +=
         "    var arg" + i + " = argType" + i + ".readValueFromPointer(args" + (offset ? ("+"+offset) : "") + ");\n";
         offset += types[i]['argPackAdvance'];
     }
     functionBody +=
-        "    var rv = handle[name](" + argsList + ");\n";
+    `    var rv = ${asCtor ? 'new func' : 'func.call'}(${argsList});\n`;
```

Comment:
indent

Reply:
Indentation in this function is generally bad - mixed with 4 and 2 spaces in different places.

I guess I can just update all at once.